src/GafferScene/ImageToPoints.cpp

```diff
@@ -0,0 +1,522 @@
+		}
+	}
+
+	// Add on our own "P" if it is not mapped from a channel.
+
+	if( !pointsPrimitive->variableData<V3fVectorData>( "P" ) )
+	{
+		V3fVectorDataPtr pData = new V3fVectorData( vector<V3f>( numPixels, V3f( 0 ) ) );
+		pointsPrimitive->variables["P"] = PrimitiveVariable( PrimitiveVariable::Vertex, pData );
+		auto it = pData->writable().begin();
+		const float pixelAspect = format.getPixelAspect();
+		for( int y = displayWindow.min.y; y < displayWindow.max.y; ++y )
+		{
+			for( int x = displayWindow.min.x; x < displayWindow.max.x; ++x )
+			{
+				*it++ = V3f( ( x + 0.5 ) * pixelAspect, y + 0.5, 0.0 );
+			}
+		}
+	}
+
+	// Add width if it is not mapped from a channel.
+
+	const float width = widthPlug()->getValue();
+	if( !pointsPrimitive->variableData<FloatVectorData>( "width" ) )
+	{
+		pointsPrimitive->variables["width"] = PrimitiveVariable( PrimitiveVariable::Constant, new FloatData( width ) );
+	}
+
+	// Compute image tiles in parallel, shuffling the data into the primitive variables
+	// via our mapping. Although we may visit tiles in any order, the ordering of points
+	// in the primitive variables is stable, based on the scanline order of the image.
+
+	ImageAlgo::parallelProcessTiles(
+		imagePlug(),
+		[&] ( const ImagePlug *image, const V2i &tileOrigin ) {
+
+			const Box2i tileBound( tileOrigin, tileOrigin + V2i( ImagePlug::tileSize() ) );
+			const Box2i validTileBound = BufferAlgo::intersection(
+				BufferAlgo::intersection( tileBound, dataWindow ), displayWindow
+			);
+
+			for( const auto &mapping : mappings )
+			{
+				ConstFloatVectorDataPtr channelData = image->channelData( mapping.name, tileOrigin );
+				const vector<float> &data = channelData->readable();
+
+				for( const auto &destination : mapping.destinations )
+				{
+					const bool isWidth = destination.name == "width";
+					const size_t stride = destination.type == FloatVectorData::staticTypeId() ? 1 : 3;
+					for( int y = validTileBound.min.y; y < validTileBound.max.y; ++y )
+					{
+						size_t inIndex = BufferAlgo::index( V2i( validTileBound.min.x, y ), tileBound );
+						size_t outIndex = BufferAlgo::index( V2i( validTileBound.min.x, y ), displayWindow ) * stride + destination.offset;
+						if( !isWidth )
+						{
+							for( int x = validTileBound.min.x; x < validTileBound.max.x; ++x )
+							{
+								destination.data[outIndex] = data[inIndex++];
+								outIndex += stride;
+							}
+						}
+						else
+						{
+							// As above, but with multiplier
+							for( int x = validTileBound.min.x; x < validTileBound.max.x; ++x )
+							{
+								destination.data[outIndex] = data[inIndex++] * width;
+								outIndex += stride;
+							}
+						}
+					}
+				}
+			}
+		},
+		BufferAlgo::intersection( displayWindow, dataWindow )
+	);
+
+	// Strip out points below the alpha threshold in a final serial step.
+	// We can't easily do this in the step above because we don't know the
+	// point count or order until after we've visited all tiles.
```

Comment:
Create large dense buffers for all channels being converted before we figure out the data layout seems pretty wasteful of performance - I strongly suspect we could do a lot better with a prepass that visits all tiles of alpha only, and then a second pass that puts all data directly into the right place.  I don't see anything that would prevent us from doing that later if we decide that performance is important for this node though.

Reply:
That approach would have very unpredictable performance when the cache is insufficient to hold all tiles for all images in the graph. Many nodes (ImageReader and all ColorProcessors to name the obvious ones) compute several channels at once, and this approach risks doing all that work twice if the top-left tiles are evicted before we begin the second pass.

The other consideration is that the serial step is actually pretty minimal compared to the typical cost of generating the input image itself. Here are some stats for the conversion of an EXR input image coming off a local SSD, straight out of an ImageReader and into an ImageToPoints node :

- ImageReader : 0.108s
- ImageToPoints : 0.123s (including ImageReader time)
- ImageToPoints with `ignoreTransparent` on : 0.133 (including ImageReader time)

So we're looking at about 8% overhead, and as image generation time increases, that overhead will only diminish. I think that's a pretty good trade-off against unpredictable performance from cache misses in more complex cases.

But there is an obvious optimisation that I missed - if `ignoreTransparent` is on, we don't need to allocate a buffer for the whole display window, because we know everything outside the data window will be discarded later. I implemented that in https://github.com/GafferHQ/gaffer/pull/5177/commits/2eed153bbe0ecfefc951e425e89a2d4201b2aa5d, for a whopping 93% reduction in runtime for a best-case scenario of a large image with a small area of text.

There are still potential optimisations for the serial step - we could do a prepass to build a vector of indices of non-zero-alpha pixels, limiting the memory accesses when doing the actual channel processing. And we could parallelize over the primitive variables. But I don't really see any need to explore that at this point.